python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -39,6 +50,56 @@
 # The minimum number of nodes to launch concurrently.
 UPSCALING_INITIAL_NUM_NODES = 5
 
+NodeResources = ResourceDict
+ResourceDemands = List[ResourceDict]
+
+
+class UtilizationScore(Protocol):
+    """This fancy class just defines the `UtilizationScore` protocol to be
+    some type that is a "totally ordered set" (i.e. things that can be sorted).
+
+    What we're really trying to express is
+
+    ```
+    UtilizationScore = TypeVar("UtilizationScore", bound=Comparable["UtilizationScore"])
+    ```
+
+    but Comparable isn't a real type and, and a bound with a type argument
+    can't be enforced (f-bounded polymorphism with contravariance). See Guido's
+    comment for more details: https://github.com/python/typing/issues/59.
```

Comment:
I'm amused that the use of "nonsense" in that comment thread is accurate, in the technical sense of "category theory."

Anyways, this class is fine.